codegen/jennies/crd.go

```diff
@@ -210,17 +210,117 @@ func CUEToCRDOpenAPI(v cue.Value, name, version string) (map[string]any, error)
-	replaceAdditionalProperties(schemaProps)
+	setKubernetesPreserveUnknownFields(schemaProps)
 
 	return schemaProps, nil
 }
 
+// setKubernetesPreserveUnknownFields replaces `"additionalProperties": {}` segments with `"x-kubernetes-preserve-unknown-fields": true`
+func setKubernetesPreserveUnknownFields(properties map[string]any) {
+	for key, val := range properties {
+		prop, ok := val.(map[string]any)
+		if !ok {
+			return
+		}
+		typ, ok := prop["type"]
+		if !ok {
+			// property has no type, not valid openAPI
+			return
+		}
+		typeString, ok := typ.(string)
+		if !ok {
+			// "type" property is not a string, not valid openAPI
+			return
+		}
+		switch strings.ToLower(typeString) {
+		case "object":
+			setKubernetesPreserveUnknownFieldsInObject(prop)
+			properties[key] = prop
+		case "array":
+			setKubernetesPreserveUnknownFieldsInArray(prop)
+			properties[key] = prop
+		}
+	}
+}
+
+func setKubernetesPreserveUnknownFieldsInObject(object map[string]any) {
+	// If the object has properties, also handle them
+	if props, ok := object["properties"]; ok {
+		if cast, ok := props.(map[string]any); ok {
+			setKubernetesPreserveUnknownFields(cast)
+			object["properties"] = cast
+		}
+	}
+	// If the object has an "additionalProperties" field, we'll need to modify the object
+	if addProps, ok := object["additionalProperties"]; ok {
+		additionalProperties, ok := addProps.(map[string]any)
+		if !ok {
+			return
+		}
+		// If additionalProperties is empty ({}), we want to delete it from `object` and replace it with "x-kubernetes-preserve-unknown-fields": true
+		if len(additionalProperties) == 0 {
+			delete(object, "additionalProperties")
+			object["x-kubernetes-preserve-unknown-fields"] = true
+		} else if innerProps, ok := additionalProperties["properties"]; ok {
+			// If "additionalProperties" contains "properties," we have to recurse and fix those properties instead
+			castInnerProps, ok := innerProps.(map[string]any)
+			if !ok {
+				return
+			}
+			setKubernetesPreserveUnknownFields(castInnerProps)
+			additionalProperties["properties"] = castInnerProps
+			object["additionalProperties"] = additionalProperties
+		}
+	}
+}
+
+func setKubernetesPreserveUnknownFieldsInArray(list map[string]any) {
+	// Get the items and check the type. If the type is "object" or "array," we have to check if it needs fixing
+	items, ok := list["items"]
+	if !ok {
+		// Not a valid array
+		return
+	}
+	castItems, ok := items.(map[string]any)
+	if !ok {
+		return
+	}
+	iType, ok := castItems["type"]
+	if !ok {
+		// No "type" in "items," not a valid array
+		return
+	}
+	itemType, ok := iType.(string)
+	if !ok {
+		// items.type must be a string
+		return
+	}
+	switch strings.ToLower(itemType) {
+	case "object":
+		setKubernetesPreserveUnknownFieldsInObject(castItems)
+		list["items"] = castItems
+	case "array":
+		setKubernetesPreserveUnknownFieldsInArray(castItems)
+		list["items"] = castItems
+	}
+}
+
+//nolint:unused
```

Comment:
I'm sorry, I can't seem to find where is this used from, could you point it out?

Reply:
It's not used, it's the old method that was used to replace the `additionalProperties`, which I fixed th ebug in, but found it to be overall a bit confusing so I replaced it with the new one. Left it in the PR in case folks thought that it was more straightforward than the new one that uses three functions instead.

Reply:
I'll remove it now